clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -643,10 +644,19 @@ private KafkaConsumer(ConsumerConfig config,
             List<ConsumerInterceptor<K, V>> interceptorList = (List) (new ConsumerConfig(userProvidedConfigs)).getConfiguredInstances(ConsumerConfig.INTERCEPTOR_CLASSES_CONFIG,
                     ConsumerInterceptor.class);
             this.interceptors = interceptorList.isEmpty() ? null : new ConsumerInterceptors<>(interceptorList);
+
+            // the maximum time we can block in a call to NetworkClient.poll is limited by the frequency
+            // of periodic events (heartbeats and autocommits)
+            boolean enableAutoCommit = config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+            int autoCommitIntervalMs = config.getInt(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG);
+            int heartbeatIntervalMs = config.getInt(ConsumerConfig.HEARTBEAT_INTERVAL_MS_CONFIG);
+            this.maxPollTimeoutMs = (enableAutoCommit ? Math.min(autoCommitIntervalMs, heartbeatIntervalMs) : heartbeatIntervalMs) / 2;
```

Comment:
Actually, the old code managed the heartbeats via DelayedTask, and that scheduling was handled internally inside poll(). But as I think about this more, I think the original code was right -- I was thinking about session timeout, but since this is for the heartbeat interval the values should be set exactly at the value such that you'll wake up and you'll be just past the heartbeat interval and send one out. Dividing by 2 doesn't help since you need to be past the heartbeat interval to send a new one out.